docs/cudf/source/errors.md

```diff
@@ -0,0 +1,46 @@
+# cuDF Error Handling Documentation (draft)
+
+## Scope
+
+This document serves to guide the use of error types when dealing with different exceptions encountered within cuDF domain. As of the draft version, it also serves as a reference for supporting libraries (like libcudf) to boost awareness of common error types encountered and to standardize their error interface. Note that this document will not serve as a complete list of error types thrown by cuDF.
+
+## Overview
+
+On a high level, exceptions consist of exception type and exception payload. Each can be standardized by the API. cuDF only standardizes exception type and uses the payload to convey debugging information for users. (Subject to discussion: is it sufficient to only use builtin error types but not the payload to signal all errors?)
+
+cuDF follows error conventions used by Pandas. Following [Pandas Wiki](https://github.com/pandas-dev/pandas/wiki/Choosing-Exceptions-to-Raise), cuDF uses python [builtin error types](https://docs.python.org/3.9/library/exceptions.html) whenever possible. Common builtin errors used in cuDF and their semantics are listed below.
+
+| builtin error type | semantic | example use case |
+| ------------------ | -------- | ---------------- |
+|     TypeError      | Operation is applied to unsupported objects. For objects that is not **yet** supported, use `NotImplementedError` | [binary operation on two unsupported column types](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/column/numerical.py#L110) |
+|     ValueError     | The value of the object does not satisfy operation requirement. Usually raising this error requires data introspection. In cases where the object only accept from a finite set of values (such as enum or string set), also use `ValueError`. | [The dataframe to describe is empty](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/dataframe.py#L5366) |
+|     IndexError     | Array access out of range | [Retrieving rows from a column specified by an out of bound index](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/column/column.py#L849-L851) |
+|     KeyError       | Mapping access with invalid key | [Retrieving rows from column specified by an invalid key](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/indexing.py#L177) |
+| NotImplementedError| Operation of object is planned, but not yet supported | (none) |
+|     IOError        | (TBA by cuIO) | (TBA) |
+
```

Comment:
Other less important errors to consider:
- `OverflowError` for numerical overflow. I believe that we recently discussed just silently allowing overflow, with the possibility of moving long-term towards the (IMO better) behavior of numpy which will at least warn (but not error) on overflow.
- Other I/O related errors (`FileNotFoundError`, `OSError`, etc). If we're rolling our own writers (or using other C++ libraries) there may be cases where we need to map the error more precisely than just `IOError`.